### PR TITLE
add logic to autoskip credits

### PR DESCRIPTION
**Pull request by phyzical**
round 2

### CODE REVIEW

**Comment by phyzical** [comment]
I tried testing this, but with the new set of changes from master the scheduled tasks to find new intros/credits no longer works?

**Comment by AbandonedCart** [comment]
If you already scanned the same content before and didn't clear the cache or select the option not to use one, then it likely had no reason to scan the same content twice.

File: ConfusedPolarBear.Plugin.IntroSkipper/AutoSkip.cs [hunk 1]
```diff
@@ -115,90 +116,95 @@ private void UserDataManager_UserDataSaved(object? sender, UserDataSaveEventArgs
...
+            adjustedStart,
+            item.IntroEnd);
 
-            if (position < adjustedStart || position > intro.IntroEnd)
-            {
-                continue;
-            }
+        if (position < adjustedStart || position > item.IntroEnd)
+        {
+            return;
```
Comment (AbandonedCart) [inline ConfusedPolarBear.Plugin.IntroSkipper/AutoSkip.cs:170]: Is all this being set to return intentionally? Continue moves to the next session. Return ends the loop entirely.

**Comment by phyzical** [comment]
oh no it takes hours to rescan then fails.

File: ConfusedPolarBear.Plugin.IntroSkipper/AutoSkip.cs (hunk 1 above)
Comment (phyzical) [reply ConfusedPolarBear.Plugin.IntroSkipper/AutoSkip.cs:170]: its now a function so the continue context happens back up at 133

**Comment by AbandonedCart** [comment]
Your PR was going against the 10.8 branch if it was attached to master

**Comment by AbandonedCart** [comment]
https://github.com/jumoog/intro-skipper/commit/5d743302e41c88f0e8ac85cee88142ca1a235aee#diff-b78d7725814906731f8885194e794164a1473bbe70e905576108bf7dc112a2c4

Those are the changes to AutoSkip from 10.8 to 10.9

**Comment by AbandonedCart** [comment]
```
using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using System.Timers;
using MediaBrowser.Common.Extensions;
using MediaBrowser.Controller.Library;
using MediaBrowser.Controller.Plugins;
using MediaBrowser.Controller.Session;
using MediaBrowser.Model.Entities;
using MediaBrowser.Model.Session;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;

namespace ConfusedPolarBear.Plugin.IntroSkipper;

/// <summary>
/// Automatically skip past introduction sequences.
/// Commands clients to seek to the end of the intro as soon as they start playing it.
/// </summary>
public class AutoSkip : IHostedService, IDisposable
{
    private readonly object _sentIntroSeekCommandLock = new();
    private readonly object _sentCreditsSeekCommandLock = new();

    private ILogger<AutoSkip> _logger;
    private IUserDataManager _userDataManager;
    private ISessionManager _sessionManager;
    private System.Timers.Timer _playbackTimer = new(1000);
    private Dictionary<string, Dictionary<string, bool>> _sentSeekCommand;

    /// <summary>
    /// Initializes a new instance of the <see cref="AutoSkip"/> class.
    /// </summary>
    /// <param name="userDataManager">User data manager.</param>
    /// <param name="sessionManager">Session manager.</param>
    /// <param name="logger">Logger.</param>
    public AutoSkip(
        IUserDataManager userDataManager,
        ISessionManager sessionManager,
        ILogger<AutoSkip> logger)
    {
        _userDataManager = userDataManager;
        _sessionManager = sessionManager;
        _logger = logger;
        _sentSeekCommand = new Dictionary<string, Dictionary<string, bool>>();
    }

    private void AutoSkipChanged(object? sender, EventArgs e)
    {
        var newState = Plugin.Instance!.Configuration.AutoSkip;
        _logger.LogDebug("Setting playback timer enabled to {NewState}", newState);
        _playbackTimer.Enabled = newState;
    }

    private void UserDataManager_UserDataSaved(object? sender, UserDataSaveEventArgs e)
    {
        var itemId = e.Item.Id;
        var newState = false;
        var episodeNumber = e.Item.IndexNumber.GetValueOrDefault(-1);

        // Ignore all events except playback start & end
        if (e.SaveReason != UserDataSaveReason.PlaybackStart && e.SaveReason != UserDataSaveReason.PlaybackFinished)
        {
            return;
        }

        // Lookup the session for this item.
        SessionInfo? session = null;

        try
        {
            foreach (var needle in _sessionManager.Sessions)
            {
                if (needle.UserId == e.UserId && needle.NowPlayingItem.Id == itemId)
                {
                    session = needle;
                    break;
                }
            }

            if (session == null)
            {
                _logger.LogInformation("Unable to find session for {Item}", itemId);
                return;
            }
        }
        catch (Exception ex) when (ex is NullReferenceException || ex is ResourceNotFoundException)
        {
            return;
        }

        // If this is the first episode in the season, and SkipFirstEpisode is false, pretend that we've already sent the seek command for this playback session.
        if (!Plugin.Instance!.Configuration.SkipFirstEpisode && episodeNumber == 1)
        {
            newState = true;
        }

        // Reset the seek command state for this device.
        lock (_sentIntroSeekCommandLock)
        {
            var device = session.DeviceId;

            _logger.LogDebug("Resetting seek command state for session {Session}", device);
            _sentSeekCommand[device]["intro"] = newState;
        }
    }

    private void PlaybackTimer_Elapsed(object? sender, ElapsedEventArgs e)
    {
        foreach (var session in _sessionManager.Sessions)
        {
            Skip(sender, e, "intro", session, _sentIntroSeekCommandLock, Plugin.Instance!.Intros, Plugin.Instance!.Configuration.AutoSkipIntroNotificationText);
            Skip(sender, e, "credit", session, _sentCreditsSeekCommandLock, Plugin.Instance!.Credits, Plugin.Instance!.Configuration.AutoSkipCreditsNotificationText);
        }
    }

    private void Skip(object? sender, ElapsedEventArgs e, string type, SessionInfo session, object type_lock, Dictionary<Guid, Intro> items, string notificationText)
    {
        var deviceId = session.DeviceId;
        var itemId = session.NowPlayingItem.Id;
        var position = session.PlayState.PositionTicks / TimeSpan.TicksPerSecond;

        // Don't send the seek command more than once in the same session.
        lock (type_lock)
        {
            if (_sentSeekCommand.TryGetValue(deviceId, out var types) && types.TryGetValue(type, out var sent) && sent)
            {
                _logger.LogTrace("Already sent intro seek command for session {Session}", deviceId);
                return;
            }
        }

        // Assert that an intro was detected for this item.
        if (!items.TryGetValue(itemId, out var item) || !item.Valid)
        {
            return;
        }

        // Seek is unreliable if called at the very start of an episode.
        var adjustedStart = Math.Max(5, item.IntroStart);

        _logger.LogTrace(
            "Playback position is {Position}, intro runs from {Start} to {End}",
            position,
            adjustedStart,
            item.IntroEnd);

        if (position < adjustedStart || position > item.IntroEnd)
        {
            return;
        }

        // Notify the user that an introduction is being skipped for them.
        if (!string.IsNullOrWhiteSpace(notificationText))
        {
            _sessionManager.SendMessageCommand(
            session.Id,
            session.Id,
            new MessageCommand()
            {
                Header = string.Empty,      // some clients require header to be a string instead of null
                Text = notificationText,
                TimeoutMs = 2000,
            },
            CancellationToken.None);
        }

        _logger.LogDebug("Sending seek command to {Session}", deviceId);

        var introEnd = (long)item.IntroEnd - Plugin.Instance!.Configuration.SecondsOfIntroToPlay;

        _sessionManager.SendPlaystateCommand(
            session.Id,
            session.Id,
            new PlaystateRequest
            {
                Command = PlaystateCommand.Seek,
                ControllingUserId = session.UserId.ToString("N"),
                SeekPositionTicks = introEnd * TimeSpan.TicksPerSecond,
            },
            CancellationToken.None);

        // Flag that we've sent the seek command so that it's not sent repeatedly
        lock (type_lock)
        {
            _logger.LogTrace("Setting seek command state for session {Session}", deviceId);
            _sentSeekCommand[deviceId][type] = true;
        }
    }

    /// <summary>
    /// Dispose.
    /// </summary>
    public void Dispose()
    {
        Dispose(true);
        GC.SuppressFinalize(this);
    }

    /// <summary>
    /// Protected dispose.
    /// </summary>
    /// <param name="disposing">Dispose.</param>
    protected virtual void Dispose(bool disposing)
    {
        if (!disposing)
        {
            return;
        }

        _playbackTimer.Stop();
        _playbackTimer.Dispose();
    }

    /// <inheritdoc />
    public Task StartAsync(CancellationToken cancellationToken)
    {
        _logger.LogDebug("Setting up automatic skipping");

        _userDataManager.UserDataSaved += UserDataManager_UserDataSaved;
        Plugin.Instance!.AutoSkipChanged += AutoSkipChanged;

        // Make the timer restart automatically and set enabled to match the configuration value.
        _playbackTimer.AutoReset = true;
        _playbackTimer.Elapsed += PlaybackTimer_Elapsed;

        AutoSkipChanged(null, EventArgs.Empty);

        return Task.CompletedTask;
    }

    /// <inheritdoc />
    public Task StopAsync(CancellationToken cancellationToken)
    {
        _userDataManager.UserDataSaved -= UserDataManager_UserDataSaved;
        return Task.CompletedTask;
    }
}
```

**Comment by phyzical** [comment]
so again ive tried 1.16.4 with 10.8 AND the beta versions for 10.9 nothing ends up skipping credits.

im talking without any of my changes nothing skips credits

**Comment by AbandonedCart** [comment]
I thought the whole purpose of this PR was to add skipping credits.

**Comment by AbandonedCart** [comment]
I ran a test with your changes and neither intro or credits skipped. Once I removed your changes, intros skipped, but credits didn't

**Comment by AbandonedCart** [comment]
For now, I made a duplicate of the AutoSkip in 10.8 and changed it to credits. It can eventually get merged together, but it is probably good to have the option to enable automation for the two separately. Hopefully it will help figure out what is going on.

https://github.com/jumoog/intro-skipper/commit/94a5a7279c744a2a58c5b0bab6bac0f7bbafa2e1

**Comment by phyzical** [comment]
Eh if you changes work ill just close this and wait for things the smooth out

**Comment by AbandonedCart** [comment]
It's only for 10.8 and it's probably about the least efficient way about it, but I wanted to test if it would even work.

**Pull request closed without merging**